### ISSUE TITLE
Migration from legacy iOS app (same Bundle ID)

### ISSUE THREAD

**Issue by michaelstingl** [owncloud/ios-app#270]
Legacy ownCloud iOS app supports `Basic Auth` and `OAuth 2.0` logins, but also has a branding option for SAML/Shibboleth login. We need to define what migrations are possible, and what happens if migration isn't possible. (No SAML/Shibboleth login in the new ownCloud iOS app)

**Comment by jesmrec** [comment]
If we do not support SAML, i guess there is no other option to migrate, so OAuth2 and basic are currently supported in the app. Not sure if you meant this.

**Comment by michaelstingl** [comment]
Yeah real migration from SAML to OAuth2 won't be possible. But we need to guide the user, so they understand whats going on.

**Comment by kleinehector** [comment]
Thanks for opening this Michael. We would be okay when a user must re-login when he opens the new app, most important is that this goes smoothly (so a user upgrades from old to new and old SAML session is silently destroyed while user is forwarded to oAuth login).

**Comment by michaelstingl** [comment]
Idea from https://github.com/owncloud/ios/issues/1109#issuecomment-505471649: 

> some inter-app-linking to send your server-url and username to the new app or request it in the new app from the old app? (maybe not the password and the data itself)

For the unbranded apps in the public App Store, I tend to not upgrade from the old app to the new one. Instead, the new app can be installed in parallel.

For branded apps (with SAML etc), we need to check how a migration could look like.

**Comment by hosy** [comment]
# Migration to new iOS App

For a seemless migration from the old iOS app to the new iOS app we should provide the user a possibility to move the existing accounts and settings to the new iOS app.

## What is needed:

- Move Accounts

## Nice to have

- Migrate Settings
  - Pin lock & Face ID
  - Upload Photos
  - Upload Videos
- Move Passwords

**Comment by CorneeldH** [comment]
@hosy Good to see an update in this issue.

At SURF we currently run iOS with SAML (no oAuth because changing from SAML to oAuth was not supported). So I second the point of Michael, because enabling moving of oAuth account from old to new app, attention for old SAML is needed: "real migration from SAML to OAuth2 won't be possible. But we need to guide the user, so they understand whats going on."

**Comment by hosy** [comment]
iOS App-Migration 

## Branded (Update with app with same App ID)

- Singleaccount / Multiaccount

- Basic Auth ==> Basic Auth
- Basic Auth ==> OAuth 2.0 (in case oauth2 enabled on server)
- OAuth 2.0 ==> OAuth 2.0
- SAML ==> OAuth 2.0

Multiaccount: Indicator that user interaction is required? (in list AND on login page with explanation)

Warning that no local data wasn't migrated

After Migration:
    Delete old file container, App database

## Unbranded ==> Inter-App Migration

- Singleaccount / Multiaccount

- Basic Auth ==> Basic Auth
- Basic Auth ==> OAuth 2.0 (in case oauth2 enabled on server)
- OAuth 2.0 ==> OAuth 2.0

Multiaccount: Indicator that user interaction is required? (in list AND on login page with explanation)

Basic Auth:
    Extended Error Message (App Passwort, 2FA => Better use OAuth 2.0)

===== ===== ===== ===== =====


Needed Data
Accounts
URL
Method
Name
Username

Recognize new iOS-App
Check via URL-Scheme?

Unlock 1 IaP with old iOS app

How to Import Data
Plist
Key-Value Sharing
Keychain-Sharing

What about different App Store accounts ??

## Settings Migration

- Check which settings can be reused from the old app to the new app
- Upload path for photos/videos
- Pass Lock / Face ID


## Migration UI

- Add a view controller to give user feedback about the migration process
- check list style with indicator and icon after the task was finished


## Migration Framework

- Check if it makes sense to build an Objective-C framework to use the code for the legacy and new iOS app
- Benefit: not needed to write two independent migration classes
- indepentent from the new iOS app SDK

-> Result: We decided to not built a framework. Migration will happens directly in the app.

## Unit-Tests

- implement test cases to automate migration

**Comment by jesmrec** [comment]
Done! Ready to go in 11.4
